tanjun/dependencies/limiters.py

```diff
@@ -147,6 +149,82 @@ async def try_acquire(self, bucket_id: str, ctx: tanjun.Context, /) -> bool:
     async def release(self, bucket_id: str, ctx: tanjun.Context, /) -> None:
         """Release a concurrency lock on a bucket."""
 
+    def acquire(
+        self,
+        bucket_id: str,
+        ctx: tanjun.Context,
+        /,
+        *,
+        error: collections.Callable[[], Exception] = lambda: errors.CommandError(
+            "This resource is currently busy; please try again later."
+        ),
+    ) -> contextlib.AbstractAsyncContextManager[None]:
+        """Acquire an concurrency lock on a bucket through an async context manager.
+
+        Parameters
+        ----------
+        bucket_id
+            The concurrency bucket to acquire.
+        ctx
+            The context to acquire this resource lock with.
+        error
+            Callback which returns the error that's raised when the lock
+            couldn't be acquired due to being at it's limit.
+
+            This will be raised on entering the returned context manager and
+            defaults to an English command error.
+
+        Returns
+        -------
+        contextlib.AbstractAsyncContextManager[None]
+            The context manager which'll acquire and release this concurrency lock.
+
+        Raises
+        ------
+        tanjun.errors.CommandError
+            The default error that's raised while entering the returned async
+            context manager if it couldn't acquire the lock.
+        """
+        return _Acquire(self, bucket_id, ctx, error)
+
+
+class _Acquire:
+    __slots__ = ("_acquired", "_bucket_id", "_ctx", "_error", "_limiter")
+
+    def __init__(
+        self,
+        limiter: AbstractConcurrencyLimiter,
+        bucket_id: str,
+        ctx: tanjun.Context,
+        error: collections.Callable[[], Exception],
```

Comment:
consider defaulting this to ResourceDepleted (and CooldownDepleted for cooldowns) instead of CommandError